embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+                Poll::Pending
+                // Note we leave the existing UnsafeCell contents - they still
+                // contain the original message. We could create another UnsafeCell
+                // with the message of Full, but there's no real need.
+            }
+        }
+    }
+}
+
+struct CloseFuture<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    sender: Sender<'ch, M, T, N>,
+}
+
+impl<'ch, M, T, const N: usize> Future for CloseFuture<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.sender.is_closed() {
+            Poll::Ready(())
+        } else {
+            self.sender
+                .channel
+                .get()
+                .set_senders_waker(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+}
+
+impl<'ch, M, T, const N: usize> Drop for Sender<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    fn drop(&mut self) {
+        self.channel.get().deregister_sender()
+    }
+}
+
+impl<'ch, M, T, const N: usize> Clone for Sender<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    #[allow(clippy::clone_double_ref)]
+    fn clone(&self) -> Self {
+        self.channel.get().register_sender();
+        Sender {
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+/// An error returned from the [`try_recv`] method.
+///
+/// [`try_recv`]: super::Receiver::try_recv
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum TryRecvError {
+    /// A message could not be received because the channel is empty.
+    Empty,
+
+    /// The message could not be received because the channel is empty and closed.
+    Closed,
+}
+
+/// Error returned by the `Sender`.
+#[derive(Debug)]
+pub struct SendError<T>(pub T);
+
+impl<T> fmt::Display for SendError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "channel closed")
+    }
+}
+
+/// This enumeration is the list of the possible error outcomes for the
+/// [try_send](super::Sender::try_send) method.
+#[derive(Debug)]
+pub enum TrySendError<T> {
+    /// The data could not be sent on the channel because the channel is
+    /// currently full and sending would require blocking.
+    Full(T),
+
+    /// The receive half of the channel was explicitly closed or has been
+    /// dropped.
+    Closed(T),
+}
+
+impl<T> fmt::Display for TrySendError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "{}",
+            match self {
+                TrySendError::Full(..) => "no available capacity",
+                TrySendError::Closed(..) => "channel closed",
+            }
+        )
+    }
+}
+
+struct ChannelState<T, const N: usize> {
+    buf: [MaybeUninit<UnsafeCell<T>>; N],
+    read_pos: usize,
+    write_pos: usize,
+    full: bool,
+    closing: bool,
```

Comment:
Why is `closing` needed? I think on close you can immediately set `closed = true`, but on recv check if there's available data anyway.

Reply:
I think what I have is more explicit. The channel can be closed explicitly prior to performing a `recv`, in which case you need to continue to consume. However, you shouldn't consume if closed...

Reply:
It's extra code and 1 extra byte of state for IMO no benefit. We should try to make it as efficient as possible.

Reply:
Here's a scenario to consider:

1. there is one sender
2. sender sends
3. the sender is dropped
4. the receiver then receives

When a sender is dropped, it will de-register from the channel. When there are no senders left, the channel is put into a `closing` state. The channel is not closed at this stage. If it was then the receiver would not be able to receive what was sent.

The `receiver_receives_given_try_send_async` test covers this scenario.

Reply:
You can still achieve that without the `closing` state. Receive does the following:
- If buffer not empty, pop a value, return it.
- Else, if closed return Closed error
- Else return Empty error

Channel is immediately closed, but receivers won't see it until after draining all the items.

Reply:
I'm convinced. :-) Commit bed15ada14959186bcb308bd981489b6d7426773 removed the `closing` flag.